### PR TITLE
Ia 499 create versions on source

### PR DESCRIPTION
**Pull request by quang-le**
Features:

- user can launch task with current DHIS2 credentials when they exist
- user can launch task with credentials other than current
- DHIS2 credentials are required when none exist
- user can choose between staying on page or go to created task when new task is created
- version number is bumped automatically
-  `force` field is always `false`
- user can define credentials in create and update modals

Other changes:
- updated `/iaso/api/tasks.py` to enable passing ordering parameters.
- added `postRequestHandler` to `/hat/assets/js/apps/Iaso/utils/requests.js`. This handler manages both success and error snack bars.
- added `/hat/assets/js/apps/Iaso/domains/dataSources/requests.js` to store domain-specific request
- added `EditableTextFields`, `UneditableTextFields` and `Checkboxes` to `/hat/assets/js/apps/Iaso/components/forms`

Points of attention:
- wording of fields and buttons
- naming conventions in code (snake case/camel case alternance) 
- the `DHIS name` field is useless ad the `dhis_ou_importer` api doesn't use it
- API error can trigger memory leak warning

https://user-images.githubusercontent.com/38907762/116383682-5fce4580-a817-11eb-8be0-581cc26b041d.mov

### CODE REVIEW

File: hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        if (showDefaultOverride) {
+            checkboxes.push({
+                keyValue: 'change_source',
+                label: MESSAGES.useDefaultDhisSettings,
+                value: withExistingDhis2Settings,
+                onChange: setWithExistingDhis2Settings,
+            });
+        }
+        return (
+            <>
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js:163]: Are you sure you need a fragment here ?

File: hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            </>
+        );
+    };
+    return (
+        <ConfirmCancelDialogComponent
+            renderTrigger={renderTrigger}
+            titleMessage={titleMessage}
+            onConfirm={onConfirm}
+            // eslint-disable-next-line no-unused-vars
+            onClosed={closeDialog => reset()}
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js:221]: `() => reset()` ?

File: hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.spec.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.spec.js:1]: 😄

File: hat/assets/js/apps/Iaso/domains/dataSources/components/DataSourceDialogComponent.js [hunk 4]
```diff
@@ -44,14 +46,24 @@ export class DataSourceDialogComponent extends Component {
...
-
         let saveCurrentDataSource;
         const currentDataSource = {};
         Object.keys(form).forEach(key => {
             if (key !== 'is_default_source') {
                 currentDataSource[key] = form[key].value;
             }
         });
+        this.setState({ hasConfirmed: true });
+        currentDataSource.credentials = {
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/dataSources/components/DataSourceDialogComponent.js:57]: i did this, maybe find a better way ...

File: hat/assets/js/apps/Iaso/domains/dataSources/config.js [hunk 5]
```diff
@@ -59,27 +61,65 @@ const dataSourcesTableColumns = (
...
+                sortedVersions.length > 0 ? sortedVersions[0].number : 0;
+            const addTaskTitle = {
+                id: 'addAskTitle',
+                defaultMessage: `${formatMessage(MESSAGES.addTask)} - Source: ${
+                    settings.original.name
+                } - Version: ${latestVersion + 1}`,
+            };
+
+            return (
+                <section>
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/dataSources/config.js:78]: Again i'm not sure the section is required here

File: iaso/api/data_sources.py [hunk 6]
```diff
@@ -43,9 +49,21 @@ def get_projects(obj: DataSource):
         return ([v.as_dict() for v in obj.projects.all()],)
 
     def create(self, validated_data):
+        # TO-DO use credentials serializer as https://github.com/BLSQ/iaso/blob/development/iaso/api/forms.py
```
Comment (beygorghor) [inline iaso/api/data_sources.py:52]: @madewulf, if you want we can create a serializer for ExternalCredentials, up to you

**Review by beygorghor — COMMENTED** [review]
Screenshot 2021-04-28 at 16 27 12
Some padding is missing here

File: hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js (hunk 2 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js:221]: maybe even `reset`. I'll test it

File: hat/assets/js/apps/Iaso/domains/dataSources/components/DataSourceDialogComponent.js (hunk 4 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/dataSources/components/DataSourceDialogComponent.js]: fixed in https://github.com/BLSQ/iaso/pull/110/commits/034d4176b8be75ff050db018cbd19d9e5c8482c8

File: hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js (hunk 2 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js]: fixed in https://github.com/BLSQ/iaso/pull/110/commits/034d4176b8be75ff050db018cbd19d9e5c8482c8

File: hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js (hunk 1 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js]: https://github.com/BLSQ/iaso/pull/110/commits/034d4176b8be75ff050db018cbd19d9e5c8482c8

Comment (madewulf) [reply hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js]: Please change the title to "import a new version from DHIS2"

Comment (madewulf) [reply hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js]: And in general, it's better to say DHIS2 than DHIS

Screenshot 2021-05-03 at 14 07 15

Comment (madewulf) [reply hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js]: Other point: you should prevent removal of the last project (there should always be one). If you remove all projects, the data source becomes dangling in the database and is not usable by anyone.

Comment (madewulf) [reply hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js]: I would also force the choice of the "default version" and when nothing is stored in the database, pick the latest one.

Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/dataSources/components/AddTaskComponent.js]: prevent removal of last project fixed in https://github.com/BLSQ/iaso/pull/110/commits/0c056856576907f40c2827370f53781c61c706db

**Pull request merged**